### ISSUE TITLE
Default case for .matchWith

### ISSUE THREAD

**Issue by JesterXL** [origamitower/folktale#139]
If you are missing a Union type in `matchWith`, you don't get an error telling you what you did wrong. If you look at the code, it's pretty simple to see, though:

```javascript
const AttackResult = union('AttackResult', {
    Hit(damage, isCritical) {
        return {damage, isCritical};
    },
    Miss(){}
}).derive(Equality);
const { Hit, Miss } = AttackResult;
const b = Hit(1, false);
b.matchWith({
    // Hit: ()=> log("we have a hit."),
    Miss: ()=> log("Miss!")
});
```

TypeError: pattern[name] is not a function
    at InternalConstructor.matchWith (/Users/jessewarden/Documents/_work_projects/get-eapi-oauth-token-master/node_modules/folktale/adt/union/union.js:67:27)

I KNOW I've seen another error somewhere where it was a page long and awesome. If you can guide me to where that kind of error handling is in the code, I can submit a PR for this. Additionally, is there any code in Folktale currently that has safe get behavior, similar to Lodash' [get](https://lodash.com/docs/4.17.4#get)?

Also, side note, is there any case for supplying a catch all like [Elixir has](https://elixir-lang.org/getting-started/case-cond-and-if.html)?

```javascript
b.matchWith({
    Hit: ()=> log("we have a hit."),
    Miss: ()=> log("Miss!"),
   _: ()=> log("Catch all.")
});
```

### Steps to reproduce

Create a union type with 2 types and make a matchWith only have 1 of the 2, it'll throw the error.

#### Expected behaviour

Just a nicer error like I've seen elsewhere indicating what you did wrong, and steps to help you mitigate it in the future.

#### Observed behaviour

Throws the error I showed above.

### Environment

(Describe the environment where the problem happens. This usually includes:

  - OS: Mac
  - JavaScript VM: Node 8
  - Folktale version: 2.0.0-rc1
)

### Additional information

None.

**Comment by JesterXL** [comment]
Doh, I see #138 mirrors this, heh! So... maybe this one could include the catch all, then to differentiate the issues.

**Comment by robotlolita** [comment]
Having a default case is interesting. If we do go there we should use a symbol instead of something like `_` to avoid collisions with any variant name the user picks.

That way we could have something like:

```js
const { union, any } = require('folktale/adt/union');

const List = union('List', {
  Empty: () => {},
  Cons: (value, rest) => ({ value, rest })
});

List.Empty().matchWith({
  Cons: (x) => 'cons',
  [any]: (x) => 'anything else'
});
```

**Comment by JesterXL** [comment]
There is no "empty" in JavaScript; $ and _ and all these symbols have different meanings to different people. Empty sounds like a great name... concerned someone will raise concerns they need "Empty" for whatever reason. Empty also reminds me of empty string (i.e. ""). maybe Undefined?

```javascript
const { union, any } = require('folktale/adt/union');

const List = union('List', {
  Undefined: () => {},
  Cons: (value, rest) => ({ value, rest })
});

List.Undefined().matchWith({
  Cons: (x) => 'cons',
  [any]: (x) => 'anything else'
});
```

Or Any?

Anyway, I'll update a separate PR to be Empty and see what y'all think.

**Comment by robotlolita** [comment]
Oh, `Empty` in this case is just the name of the variant for the empty list (so a list could be `Cons(1, Cons(2, Empty()))` for `[1, 2]`). I wasn't suggesting it to be used as the name of the default case.

Rather it would be more interesting for the union module to export a symbol, and for matchWith objects to use that symbol:

```js
// union.js
const Any = Symbol.for('@@folktale:adt:default');
// (...)
Variant.prototype.matchWith = function(p) {
  if (tag in p) { return p[tag] }
  else if (Any in p) { return p[Any] }
  else { throw new Error(...) }
};
// (...)
module.exports = {
  union,
  any: Any,
  ...
}
```
```js
// list.js
const { union, any } = require('folktale/adt/union'); 
const List = union('List', {
  Empty: () => {},
  Cons: (value, rest) => ({ value, rest })
});

List.Cons(1, List.Empty()).matchWith({
  Cons: (x) => 'cons', 
  [any]: (x) => 'anything else'
});
--



```

**Comment by JesterXL** [comment]
Oops, I missed the [any] part, sorry. Coffee is better than beer at reading this. Ok, got it! The symbol makes a lot sense, too.

**Comment by robotlolita** [comment]
This has been released as `folktale@2.3.0`.
